enterprise/cypher/cypher/src/main/scala/org/neo4j/cypher/internal/compatibility/v3_4/runtime/compiled/codegen/ir/expressions/Literal.scala

```diff
@@ -28,15 +28,22 @@ case class Literal(value: Object) extends CodeGenExpression {
   override def init[E](generator: MethodStructure[E])(implicit context: CodeGenContext) = {}
 
   override def generateExpression[E](structure: MethodStructure[E])(implicit context: CodeGenContext) = {
+    // When the literal value comes from the AST it should already have been converted
     val convertedValue = value match {
       case n: java.lang.Byte => n.longValue()
       case n: java.lang.Short => n.longValue()
-      case n: java.lang.Character => n.toLong
+      case n: java.lang.Character => n.toString
       case n: java.lang.Integer => n.longValue()
       case n: java.lang.Float => n.doubleValue()
       case _ => value
     }
-    structure.constantExpression(convertedValue.asInstanceOf[AnyRef])
+    val ct = codeGenType
+    if (value == null)
+      structure.noValue()
+    else if (ct.isPrimitive)
+      structure.constantExpression(convertedValue.asInstanceOf[AnyRef])
+    else
+      structure.constantValueExpression(value, ct)
```

Comment:
This is the actual fix. The `CodeGenType` was derived as `ValueType` even though it was not converted to `Value`. In practice most of the literals except `null` will be replaced with parameters by literal replacement, so it surfaced as a problem with null.